src/EventStore.Core/DataStructures/ProbabilisticFilter/MemoryMappedFileBloomFilter/BloomFilterIntegrity.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using EventStore.Common.Utils;
+using EventStore.Core.Index.Hashes;
+using Serilog;
+
+namespace EventStore.Core.DataStructures.ProbabilisticFilter.MemoryMappedFileBloomFilter {
+	// This class is responsible for maintaining and validating a 4 byte hash at the end
+	// of the cache line.
+	public class BloomFilterIntegrity {
+		public const int CacheLineSize = 64;
+		public const int HashSize = 4;
+		private const int UintHashIndex = (CacheLineSize - HashSize) / HashSize;
+		static readonly ILogger Log = Serilog.Log.ForContext<BloomFilterIntegrity>();
+
+		private static readonly IHasher _hasher = new XXHashUnsafe(0xc58f1a7b);
+
+		public static void WriteHash(Span<byte> cacheLine) {
+			Ensure.Equal(CacheLineSize, cacheLine.Length, "cacheline length");
+
+			if (IsAllOnes(cacheLine)) {
+				// corrupted line made safe, dont make it valid again by writing the hash
+				return;
+			}
+
+			var hash = _hasher.Hash(cacheLine[..^HashSize]);
+			var uints = MemoryMarshal.Cast<byte, uint>(cacheLine);
+			uints[UintHashIndex] = hash;
+		}
+
+		// If the hash is correct, return true.
+		// If the hash is not correct, but can be corrected then correct it and return true.
+		// If the hash is not correct and cannot be corrected then set all the bits in the
+		// cacheline and return false. The data is made safe for the bloom filter.
+		public static bool ValidateHash(Span<byte> cacheLine) {
+			Ensure.Equal(CacheLineSize, cacheLine.Length, "cacheline length");
+
+			if (IsAllOnes(cacheLine)) {
+				// all bytes 0xFFs. Corrupted data that has been made safe.
+				return false;
+			}
+
+			if (IsAllZeros(cacheLine)) {
+				// all bytes 0x00s. Empty data
+				return true;
+			}
+
+			// check the hash
+			var cacheToHash = cacheLine[..^HashSize];
+			var hash = _hasher.Hash(cacheToHash);
+			var targetHash = MemoryMarshal.Cast<byte, uint>(cacheLine)[UintHashIndex];
+			if (hash == targetHash)
+				return true;
+
+			// corrupt, try to recover.
+			if (CanRehash(cacheToHash, targetHash)) {
+				WriteHash(cacheLine);
+				Log.Verbose("Corrected a hash");
+				return true;
+			}
+
+			// corrupt and could not recover, set all bytes to 0xFFs to make safe
+			var ulongs = MemoryMarshal.Cast<byte, ulong>(cacheLine);
+			for (int i = 0; i < ulongs.Length; i++) {
+				ulongs[i] = ulong.MaxValue;
+			}
+
+			return false;
+		}
+
+		// The hash can be wrong simply because the flush happened between when
```

Comment:
This also means that a genuinely corrupted bit (flipped from 0 to 1) will be accepted, but I guess it should be fine since the bloom filter will still be safe and only potentially cause more false positives.

Reply:
And it's still better than making it all ones

Reply:
Yep exactly